snuba/datasets/generic_metrics_processor.py

```diff
@@ -86,24 +98,40 @@ def process_message(
         timestamp = _ensure_valid_date(datetime.utcfromtimestamp(message["timestamp"]))
         assert timestamp is not None, "Invalid timestamp"
 
+        try:
+            retention_days = enforce_retention(message["retention_days"], timestamp)
+        except EventTooOld:
+            return None
+
         keys = []
-        indexed_values = []
+        indexed_values: MutableSequence[int] = []
+        raw_values: MutableSequence[str] = []
         tags = message["tags"]
         assert isinstance(tags, Mapping), "Invalid tags type"
 
         sorted_tag_items = sorted(tags.items())
         for key, value in sorted_tag_items:
-            assert key.isdigit() and isinstance(value, int), "Tag key/value invalid"
-            keys.append(int(key))
-            indexed_values.append(value)
+            assert key.isdigit() and (
+                isinstance(value, int) or isinstance(value, str)
+            ), "Tag key/value invalid"
 
-        try:
-            retention_days = enforce_retention(message["retention_days"], timestamp)
-        except EventTooOld:
-            return None
+            keys.append(int(key))
+            if isinstance(value, int):
+                indexed_values.append(value)
+                raw_values.append("")
+            elif isinstance(value, str):
+                indexed_values.append(0)
```

Comment:
Technically 0 is a valid value. It would be assigned to the first value we indexed. Though probably not a big deal as, if we are here, we are not trusting indexed tag values anymore.

Reply:
i think there's two reasons this is ok

* https://github.com/getsentry/sentry/blob/master/src/sentry/migrations/0284_metrics_indexer_alter_seq.py

* 0 is already used as a sentinel value returned by clickhouse for the group in `group by tags[foo]` where `foo` does not exist as a tag

Reply:
One counter argument to that approach. 
Defensive design would prompt us not to rely on that sequence migration in postgres or on the fact we use the value on clickhouse as a sentinel to consider this ok. 
Those two are implementation details are not reflected in the design:
- The mutations is totally specific to postgres, it is not written anywhere that the sequence has to start from 1. So relying on that would be pretty dangerous as we create an implicit assumption that would be easy to forget. 
- Using 0 as a sentinel value is basically a bug, and it relies itself on a very weird and coutner intuitive clickhouse behavior that is not documented (returning 0 or empty string when looking for an entry that does not exists). Clickhouse can change the way it works.

Relying on the fact that the tag values column is meaningless, thus not queried, when not in indexed mode is more solid because it relies on a property of the data model design: either we query indexed values or we query non indexed values, we do not support hybrid queries. If we wanted to support hybrid queries (that use indexed values where the values are indexed and use string where values are not indexed) we would likely add an `index_state` column in the clickhosue table and rely on that one to decide if the row has indexed tags or not. 
That way we would be relying on something that is specifically designed to tell us whether the row is indexed. Thus we would be relying on something more robust than a property that happens to be true for indexed rows but does not really have to. This would appear evident to somebody new to the code base when the tribal knowledge is lost.